### DISCUSSION TITLE
ReactiveUI Routing, IViewLocator, Reactive Best Practices?

### DISCUSSION THREAD

**Discussion by bsampica** [AvaloniaUI/Avalonia#12644] [General]
I didn't really know what to title this.  I'll give you guys the basic scenario: 

Here is the hierarchy of instantiation as I understand it
`
MainWindowViewModel.cs -> MainWindow.axaml ->
HomePageViewModel.cs => HomePageView.axaml 
`
This results in a screen that has a left navigation, and a center content area, currently represented by tiles. (See Screenshot)
<img width="200" alt="Dashboard View" src="https://github.com/AvaloniaUI/Avalonia/assets/16546166/252f8cf8-2075-4b8f-baec-aa122f32e0b5">
<img width="200" alt="Operations View" src="https://github.com/AvaloniaUI/Avalonia/assets/16546166/477add2d-0947-4269-823e-c704d8e83608">

It uses the ReactiveUI RoutedViewHost to Conditionally load and transition to content inside the RoutedViewHost. 

`
<reactiveUi:RoutedViewHost. x:Name="RoutedViewHost" Router="{Binding Router}">
</reactiveUi:RoutedViewHost>
`

Ok, so at that point, I've got a navigation command connected to each of the Tiles, and the code for it is in the HomePageViewModel.cs  

```
public class HomePageViewModel : PageViewModelBase
{
    private RoutingState Router { get; } = Locator.Current.GetService<RoutingState>()!;
    public sealed override string UrlPathSegment { get => "HomePageView"; }

    public ReactiveCommand<Unit, IRoutableViewModel> NavigateToTankDetails
    {
        get =>
            ReactiveCommand.CreateFromObservable(() =>
                Router.Navigate.Execute(new TankDetailsViewModel()));
    }
}
```

Here's where it gets tricky.  I don't think the Router is actually being pulled out of the Locator, because an exception is being swallowed somewhere.  I'm not making it to the TankDetails view.  

I guess I'm looking for some advice I how I should be doing this.  It seem like it's going to get VERY messy fast, to have view models handling routing on their own, especially in cases of clicking a top level link, then wanting to use a back button to navigate back down into a hierarchy.  

At the TankDetailsViewModel and View there would be a back button that would take the user back to the previous page (however they got there) using the Router.NavigateBack.Execute() - but again, that means I need a reference to the router, or a different way to do this.

Most of my Splat stuff is being loaded up in the MainWindowViewModel.cs as in below

```
public MainWindowViewModel()
    {
        Router = new RoutingState();
        Locator.CurrentMutable.RegisterViewsForViewModels(Assembly.GetCallingAssembly());
        Locator.CurrentMutable.RegisterLazySingleton(() => Router);

        NavigateToOperations = ReactiveCommand.CreateFromObservable(() =>
            Router.Navigate.Execute(new OperationsViewModel()));

        NavigateToHomeScreen = ReactiveCommand.CreateFromObservable(() =>
            Router.Navigate.Execute(new HomePageViewModel()));
    }
```

All that appears to work well, but is the MainWindow the right spot to have the wire up of all the IOC DI?  

Thank you all for any advice anyone can give me.   Here's the short version::

- Where is the best place to wire up the IOC DI and have it be available?  Program.cs?  App.cs?
- When using the ReactiveUI Router, is there a best practices on setting and consuming it, and wiring navigation commands and routes?
- Why would my locator fail in the case of above, trying to get a reference to the routing state stored in the DI Container?

Github repo available for view here:
https://github.com/bsampica/VibeOne

**Comment by bsampica** [comment]
Ok, I was able to get it working as is.  The exception being thrown down low was my own fault.  So generally this pattern, or rather anti-pattern works, but I'm still interested in the best practices side of things, and how the recommended structure is.

I'm editing the title of this to reflect the question better now that the exception is resolved.

**Comment by timunie** [comment]
You can also implement IDataTemplate to have a better control over your DataTemplate-selection, for example:
https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/Routing/BasicViewLocatorSample

**Reply by bsampica** [reply]
Thanks @timunie - 
The Reactive MVVM template starts with the IViewLocator inheriting from IDataTemplate, is that generally the recommended way of managing multiple views?  

I’ve been through that sample before, and it seemed like when implementing the router through reactive, I was running into trouble with it trying to store / figure out types in the routing state…but I’ll take  another look at it.  Should I just keep this discussion going or start a new one with my findings?

**Reply by timunie** [reply]
In 0.10.x it was. Now in 11.0 it is not there by default. Often plain DataTemplates matching your DataType are enough for Avalonia.

**Reply by bsampica** [reply]
I just created a brand new 11.0 project, and the ViewLocator for sure inherits from the IDataTemplate still...am I not understanding the point you're trying to make?  It wouldn't be the first time, especially when I've been coding a lot.  :)

```
public class ViewLocator : IDataTemplate
{
    public Control Build(object data)
    {
        var name = data.GetType().FullName!.Replace("ViewModel", "View");
        var type = Type.GetType(name);

        if (type != null)
        {
            return (Control)Activator.CreateInstance(type)!;
        }
        
        return new TextBlock { Text = "Not Found: " + name };
    }

    public bool Match(object data)
    {
        return data is ViewModelBase;
    }
}

**Reply by timunie** [reply]
it uses Reflection which has some draw backs. That's why we reduced usage of this. Still works for sure.

**Comment by bsampica** [comment]
Thanks, I appreciate the follow up. I don’t like the reflection much, especially because of my intended deployment to Raspberry Pi.

Is there a full example of the best practices way of doing it?  Or is that what you linked?   Using the IDataTempate is the recommended way now.  If that’s the case I’ll architect a solution.

Thanks again!

**Reply by timunie** [reply]
https://docs.avaloniaui.net/docs/next/concepts/templates/data-templates-collection 

In fact you have many solutions in Avalonia. Just use the DataTemplates in any node of your XAML, implement IDataTemplate, use a FuncDataTemplate....

Explore also: https://github.com/AvaloniaUI/Avalonia.Samples#%EF%B8%8F-datatemplate-samples

**Reply by robloo** [reply]
Or write your own ViewLocator that requires registering the view/viewmodel associations ahead of time. That's type-safe and reflection-free and also allows differing naming conventions in the edge cases.

Prism can be used for ideas: https://prismlibrary.com/docs/viewmodel-locator.html#custom-viewmodel-registrations.

**Reply by stevemonaco** [reply]
I have some [AOT-friendly ViewLocator minimal samples](https://github.com/stevemonaco/AvaloniaViewModelFirstDemos), but you need to be prepared to dive into sourcegen DI approaches, some experimental.

**Reply by bsampica** [reply]
@stevemonaco - That's an excellent repo.  You can bet I'll be studying your use of it for inspiration over the next couple of days.  I come from the ASP.NET and .NET Core world, so AOT and Source Generators are not big in my wheel house.  Us web guys just throw it all into a server memory pump and laugh at startup times.  😄 

@robloo - Does Prism work well in Avalonia?  I'm not unfamiliar with Prism from back in my WPF / Silverlight and Desktop app days...and we loved it when we've used it previously.  I'd sure like to look at it as a resource.  Thanks for the link.

@timunie  - I've seen most of those samples in that repository, but I haven't connected the dots yet, I might need a little more hand holding to get there. 😄 

Thanks everyone that has jumped in on this.  I'm going to go bang out some more code and come back when I get stuck again.

**Reply by bsampica** [reply]
Sorry...what is preferred, if I start a new topic, or keep this thread going?  I feel like I've gotten some pretty good answers to my question, and the main answer is "how do I want it to work" - that there isn't a 100% - do it this way answer.  So it can be closed as successfully discussed and answered. 👏

**Reply by robloo** [reply]
@bsampica 

> Does Prism work well in Avalonia? I'm not unfamiliar with Prism from back in my WPF / Silverlight and Desktop app days...and we loved it when we've used it previously. I'd sure like to look at it as a resource. Thanks for the link.

There is a port of Prism for Avalonia here: https://github.com/AvaloniaCommunity/Prism.Avalonia. I think it's fairly well supported but you might want to talk with @DamianSuess over in that repo. I have no experience with it on my end.

**Reply by DamianSuess** [reply]
Thank you @robloo for the head-nod
.
I can't deny it, [Prism.Avalonia](https://github.com/AvaloniaCommunity/Prism.Avalonia) is my go-to approach for solving MVVM navigation and modular woes (_sorry, shamless plug_) 😉

@bsampica, to assist with your inquiry, I'm updating the [Learn.PrismLibrary](https://github.com/DamianSuess/Learn.PrismLibrary) repository to include navigating with the `SplitView` control. This may assist with giving you a working example of navigating similar to what you're trying to accomplish. Just a heads up, I still gotta wrap up the collapsing of the sidebar pane later this evening. I whipped up this sample over my lunch break.

Btw, when you get your project up and running, I'd like to give it a shot in my setup 👍

**Reply by DamianSuess** [reply]
@bsampica @robloo
For what it's worth, I wrapped up a SplitView example in the [Learn.PrismLibrary](https://github.com/DamianSuess/Learn.PrismLibrary) (_Avalonia-Ex3-SplitView_) repository showing how to collapse and expand the sidebar. This was a decent exercise and added value for future samples.

Just don't tell my wife I was coding instead of painting the bathroom on a Friday evening. 🤫

**Reply by bsampica** [reply]
Hey @DamianSuess - I was just going back through this message thread, and digging into the links and setting up my overnight reading list...and the Prism.Avalonia stuff in your repository came up.  This is amazing work, and I'm sooooo happy to see that Prism is being subset into projects like this, through individuals like you.  It's now a strong consideration for my project, and you and I will probably talk soon if you have some moments.

**Comment by bsampica** [comment]
Ok, sorry again...one last question:

It appears from examples in this thread, that most of the patterns I see preferred by you guys are IDataTemplate implementations.  That seems to be the preferred method?

I kind of like that approach, because it seems a little more generic, and doesn't require the routedviewhost as the single place of entry.  So for each of you personally, do you prefer your user controls to be IDataTemplates, or ReactiveObjects or IRoutableViewModels?

**Reply by stevemonaco** [reply]
I don't have particularly complex routing, so I prefer being able to assign a VM to a property and having navigation happen as a side effect. You can achieve this by using `<ContentControl Content="{Binding ActiveContent}" />` where `ActiveContent` is a ViewModel property. Changes will trigger Avalonia to look up what View should be displayed for the particular ViewModel type or state.

Two main approaches to help Avalonia do the View lookup: 1. create a ViewLocator (ie. `IDataTemplate`) and resolve views from the type via naming convention (reflection) or by registered factory like in my AOT sample. 2. Create a `DataTemplate` in XAML to associate every ViewModel with a particular View like below (you could also place the entire View's contents there instead of making a `UserControl`):

```xml
<DataTemplate x:DataType="vm:SomeViewModel">
    <v:SomeView />
</DataTemplate>
```

I prefer having less in XAML (harder to debug, more silent failing), so I primarily use `IDataTemplate` to resolve Views. I'll use inline`<DataTemplate>`s when the display is simple or when the type has more than one View representation.

**Reply by timunie** [reply]
> It appears from examples in this thread, that most of the patterns I see preferred by you guys are IDataTemplate implementations.

That's what I prefer. But everyone has a different use-case, so there is no one size fit all. If you have a small app and a fairly small team, I'd go for this approach. But if you develop something like the MS Office solution, maybe a different approach is needed.

**Comment by bsampica** [comment]
> I prefer having less in XAML (harder to debug, more silent failing), so I primarily use IDataTemplate to resolve Views. I'll use inline<DataTemplate>s when the display is simple or when the type has more than one View representation.

This...this all day every day.  I'm a huge fan of less XAML for anything that doesn't directly represent display logic.  Thanks for the samples.

**Comment by bsampica** [comment]
> Thank you @robloo for the head-nod . I can't deny it, [Prism.Avalonia](https://github.com/AvaloniaCommunity/Prism.Avalonia) is my go-to approach for solving MVVM navigation and modular woes (_sorry, shamless plug_) 😉
> 
> @bsampica, to assist with your inquiry, I'm updating the [Learn.PrismLibrary](https://github.com/DamianSuess/Learn.PrismLibrary) repository to include navigating with the `SplitView` control. This may assist with giving you a working example of navigating similar to what you're trying to accomplish. Just a heads up, I still gotta wrap up the collapsing of the sidebar pane later this evening. I whipped up this sample over my lunch break.
> 
> Btw, when you get your project up and running, I'd like to give it a shot in my setup 👍

Thanks man!  I look forward to seeing what you come up with.  👍   As a self-plug.  Im decently happy with the sidenav / Split View setup I threw together, with the expandable sidebar.  It's not generic enough yet, and relies on hard xaml for items.  I'd like it to be data driven, which is easy enough when I'm ready.  

If you clone my repo, it's VERY in flux, but the side menu normally keeps mostly working.   

You mean you'd like to try it on your brewery setup?  That would be freaking amazing actually.  Currently we support Raspberry PI 4 and 4b officially, and can work with instrumentation for temp, pressure, current, TDS, level sense, carbonization (through C02 saturation), generic relays, run time on pumps and motors,  and a few other things related specially to the process, like evaporation, and some curves related to aeration and temp stuff.  Most hardware that can work with 2 wire, Serial Com (UART), 4-20ma, or I2C and SPI.  It's a LOT of working hardware so setups can be really custom and still programable.
